### ISSUE TITLE
FP16 x gradient clipping

### ISSUE THREAD

**Issue by ZhaofengWu** [PyTorchLightning/pytorch-lightning#1381]
According to [apex docs](https://nvidia.github.io/apex/advanced.html), we should have a separate case for gradient clipping under fp16. But it seems not to be the case in pytorch-lightning. Is that correct? https://github.com/PyTorchLightning/pytorch-lightning/blob/732eaee4d735dafd1c90728e5583341d75ff72b5/pytorch_lightning/trainer/training_tricks.py#L26

**Comment by Borda** [comment]
@neggert @tullie pls ^^

**Comment by tullie** [comment]
Should we just fix this when we integrate pytorch mixed precision training https://github.com/PyTorchLightning/pytorch-lightning/issues/1337#issue-592324058?

**Comment by Borda** [comment]
depends on what will be faster lol

**Comment by williamFalcon** [comment]
yeah, let's fix with pytorch mixed precision. We need to remove nvidia apex

**Comment by mcarilli** [comment]
https://pytorch.org/docs/master/notes/amp_examples.html#gradient-clipping

**Comment by samgd** [comment]
This is still an issue in 0.7.5 and it's quite critical for training certain recurrent models where mixed precision can cause a 10x reduction in training time. 

Can anyone confirm that this is going to be in the 0.7.6 release? We've been hoping it makes 0.7.3, 0.7.4, 0.7.5...

**Comment by williamFalcon** [comment]
@samgd we already added this! you need to install the latest pytorch from nightly or a specific commit and lightning will use native amp. 

it’s not us you’re waiting for haha, it’s pytorch 1.6. We’re already compatible with native amp. @mcarilli

**Comment by samgd** [comment]
Okay, thanks. 

Just to confirm this _won't_ work with 0.7.5 and PyTorch 1.5? What does PyTorch 1.6 add over 1.5, I thought the latter already has native amp?

**Comment by williamFalcon** [comment]
1.6 does not use nvidia apex. it uses its internal version

**Comment by samgd** [comment]
Awesome, looking forward to it!

**Comment by Borda** [comment]
> @samgd we already added this! you need to install the latest pytorch from nightly or a specific commit and lightning will use native amp.
> 
> it’s not us you’re waiting for haha, it’s pytorch 1.6. We’re already compatible with native amp. @mcarilli

maybe we shall consider to fix it also for lower versions as we still keep back-compatibility...

**Comment by mcarilli** [comment]
[`unscale_` ](https://github.com/PyTorchLightning/pytorch-lightning/blob/2950f669834506f8e5845b318b0f25d52d19e331/pytorch_lightning/trainer/training_loop.py#L617) before clipping was added in the native amp PR, good.  But it might need to be more complicated if there are multiple models and/or optimizers.  Does the call to [`clip_gradients()`](https://github.com/PyTorchLightning/pytorch-lightning/blob/2950f669834506f8e5845b318b0f25d52d19e331/pytorch_lightning/trainer/training_loop.py#L618) clip just the gradients owned by `optimizer` at that point, all the gradients in the network, or the gradients owned by a particular model?

**Comment by edenlightning** [comment]
@mcarilli mind checking if the issue is still on master?

**Comment by edenlightning** [comment]
@Borda still relevant?

**Comment by teddykoker** [comment]
Verified clipping works with native fp16

**Comment by williamFalcon** [comment]
@teddykoker do we have a test to prevent regressions?

**Comment by teddykoker** [comment]
Don't believe so, I can add one

**Comment by williamFalcon** [comment]
it it's quick.
